rfc/plugins-ui.md

```diff
@@ -1,115 +1,124 @@
 
-### Loading a Plugin
-After Deck queries Front50 for enabled plugins, it gathers the metadata necessary to load the enabled plugins. Deck will then use native module loading to import all required plugins. 
+## Bootstrapping Deck
+Plugins will be able to use shared library code which is provided by Deck itself (see [Plugin Structure](#plugin-structure) below).  These shared libraries should be loaded before the plugins themselves.  Because of this, the bootstrapping process must change.  The order of operations likely should be:
 
-Downloading of plugin resources occurs before the app boots (Check out Bootstrapping Deck to learn more). Once those resources are downloaded, plugins are initialized as described below. We accomplish this by moving the loading of the netflix.spinnaker module to a finally block after initializing the plugins. This ensures that it is always called. This is seen below:
+- Load Deck assets
+- Load plugin manifest
+- Load plugin assets
+- Deck Bootstrap
+- Plugin Initialization
+- Start Deck
+
+####  Deck Bootstrap
+Today, Deck is bootstrapped via AngularJS as soon as all the declared AngularJS modules dependencies and available.  We have to delay starting AngularJS while the plugins load. In order to delay the starting of AngularJS, we have to manually bootstrap the application after all the plugins are loaded.  A small snippet such as the this pseudocode can be added to `index.html`.
 
 ```
-loadPlugins()
-  .catch(() => {
-    // deal with errors
-  })
-  .finally(() => {
-    // load netflix.spinnaker module
-  }
-
-let loadedPlugins = []
-
-export function loadPlugins() {
-  const plugins = fetch from Front50;
-  return Promise.all(plugins.map(
-		// import each plugin and add it to the loadedPlugins
-		//   if a plugin fails to load, log it
-  ));
-}
+// loadPluginsPromises is an array of Promise
+Promise.all(loadPluginsPromises).then(loadedPlugins => {
+  // Tell AngularJS to start
+  bootstrap(document.documentElement, ['netflix.spinnaker']);
+}).catch(error => {
+  // Handle plugin loading errors
+});
 ```
 
-### initPlugins
-After all of the plugins have been loaded via import, they need to be initialized. It will be ideal to have plugins not need to know about Deck internals to register themselves.  
-
-Stop Gap Solution for Registering Plugins:
-At the very beginning there will not be well defined extension points for everything. To get around this, plugins may have to register themselves until a well defined extension point is created.
+### Plugin Initialization
 
-Well Defined Extension Points:
-Interfaces need to be created for anything that can be extended. 
+During Deck's bootstrap, it should call all enabled plugins' `initialize()` method passing in a Plugin API object.  This object will expose a set of hand-picked functions that are hand picked for Spinnaker Core and register.
```

Comment:
If I'm reading this correctly, plugins will still have to register themselves - did you see any way of using an IoC pattern here? From the design doc, we discussed possibly having Deck register plugins it finds (https://docs.google.com/document/d/16WmRSziTJsSBZ1kuKUfVleLAYMIxmfvz/edit?disco=AAAAD8PetDs).

Second sentence has some redundant info, hand-picked / hand picked functions

Reply:
I will update this. 

My main concern is that some things in deck (especially related to routing) may not be able to be done declaratively and will still require an imperative API such as `initialize()`.  However, I agree it would be better to export an object with (for example) a `stages` property than have the initialize method wire them up.